src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/DefaultRazorCompletionFactsService.cs

```diff
@@ -4,97 +4,42 @@
 using System;
 using System.Collections.Generic;
 using System.Composition;
-using System.Linq;
 using Microsoft.AspNetCore.Razor.Language;
-using Microsoft.AspNetCore.Razor.Language.Legacy;
-using Microsoft.AspNetCore.Razor.Language.Syntax;
 
 namespace Microsoft.CodeAnalysis.Razor.Completion
 {
     [Shared]
     [Export(typeof(RazorCompletionFactsService))]
     internal class DefaultRazorCompletionFactsService : RazorCompletionFactsService
     {
-        private static readonly IEnumerable<DirectiveDescriptor> DefaultDirectives = new[]
-        {
-            CSharpCodeParser.AddTagHelperDirectiveDescriptor,
-            CSharpCodeParser.RemoveTagHelperDirectiveDescriptor,
-            CSharpCodeParser.TagHelperPrefixDirectiveDescriptor,
-        };
-
-        public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorCodeDocument codeDocument, SourceSpan location)
-        {
-            if (codeDocument is null)
-            {
-                throw new ArgumentNullException(nameof(codeDocument));
-            }
-
-            var completionItems = new List<RazorCompletionItem>();
-            var syntaxTree = codeDocument.GetSyntaxTree();
-
-            if (AtDirectiveCompletionPoint(syntaxTree, location))
-            {
-                var directiveCompletions = GetDirectiveCompletionItems(syntaxTree);
-                completionItems.AddRange(directiveCompletions);
-            }
+        private readonly IEnumerable<RazorCompletionItemProvider> _completionItemProviders;
 
-            return completionItems;
-        }
-
-        // Internal for testing
-        internal static List<RazorCompletionItem> GetDirectiveCompletionItems(RazorSyntaxTree syntaxTree)
+        [ImportingConstructor]
+        public DefaultRazorCompletionFactsService([ImportMany] IEnumerable<RazorCompletionItemProvider> completionItemProviders)
```

Comment:
I think splitting this up is probably a good change overall. What's the plan to actually make the set of providers configurable? TBD?

Reply:
My current plan is for it to be configurable by omission. For instance, VS windows/ mac wont have all of VSCodes TagHelper completion providers so they wont end up conflicting with Todds HTML TagHelper completion. Howeverr, in VSCode world I'll register them as providers to enable them to flow through the same systems. Once we make the move to consume VSCode's TagHelper completion bits it'll be as simple as copying over the classes into Razor.Workspaces since the classes will already be using Razor's infrastructure.

My future what-if plan for if we end up moving faster in VSWindows than VSMac is to add MEF metadata support to restrict platforms OR i'll utilize shared files to have the TagHelper completions in the VSCode and VSWindows specific worlds but not mac. Basically, with this provider model anything be possible.

Reply:
ok sounds good. I wasn't sure if we needed something contextual or not. Per-platform will be easy to manage.